resilience4j-bulkhead/src/main/java/io/github/resilience4j/bulkhead/adaptive/AdaptiveBulkhead.java

```diff
@@ -0,0 +1,546 @@
+				throw e;
+			} finally {
+				if (start != null) {
+					Instant finish = Instant.now();
+					bulkhead.onComplete(Duration.between(start, finish), true);
+				}
+			}
+		};
+	}
+
+	static void handleError(AdaptiveBulkhead bulkhead, @Nullable Instant start, Exception e) {
+		if (bulkhead.getBulkheadConfig().getAdaptIfError() != null) {
+			//noinspection unchecked
+			if (bulkhead.getBulkheadConfig().getAdaptIfError().test(e) && start != null) {
+				Instant finish = Instant.now();
+				bulkhead.onComplete(Duration.between(start, finish), false);
+
+			}
+		} else {
+			bulkhead.releasePermission();
+		}
+	}
+
+	/**
+	 * Create a Bulkhead with a default configuration.
+	 *
+	 * @param name the name of the bulkhead
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead ofDefaults(String name) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, AdaptiveBulkheadConfig.ofDefaults());
+	}
+
+
+	/**
+	 * Creates a bulkhead with a custom configuration and custom limiter
+	 *
+	 * @param name         the name of the bulkhead
+	 * @param config       a custom BulkheadConfig configuration
+	 * @param limitAdapter the custom limit adopter
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config, LimitPolicy<Bulkhead> limitAdapter) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config, limitAdapter);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name   the name of the bulkhead
+	 * @param config a custom BulkheadConfig configuration
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, AdaptiveBulkheadConfig config) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, config);
+	}
+
+	/**
+	 * Creates a bulkhead with a custom configuration
+	 *
+	 * @param name                   the name of the bulkhead
+	 * @param bulkheadConfigSupplier custom configuration supplier
+	 * @return a Bulkhead instance
+	 */
+	static AdaptiveBulkhead of(String name, Supplier<AdaptiveBulkheadConfig> bulkheadConfigSupplier) {
+		return AdaptiveLimitBulkhead.factory().createAdaptiveLimitBulkhead(name, bulkheadConfigSupplier.get());
+	}
+
+	interface AdaptiveBulkheadMetrics extends Bulkhead.Metrics {
+
+		/**
+		 * @return average latency for service calls in millis
+		 */
+		double getAverageLatencyMillis();
+
+		/**
+		 * @return current failure rate for recorded calls if error check is enabled for related exceptions
+		 */
+		double getFailureRate();
+
+		/**
+		 * @return current slow call rate for the recorded calls
+		 */
+		double getSlowCallRate();
+	}
+
+	/**
+	 * An EventPublisher which can be used to register event consumers.
+	 */
+	interface AdaptiveEventPublisher extends io.github.resilience4j.core.EventPublisher<BulkheadLimit> {
+
+		EventPublisher onLimitIncreased(EventConsumer<BulkheadOnLimitDecreasedEvent> eventConsumer);
```

Comment:
We should also add `onSuccess`, `onError`, `onIgnoredError`, 'onCallNotPermitted'  and `onStateTransition`.  
We should keep a state in the AdaptiveBulkhead implementation and report state changes from:  Slow Start -> Congestion Avoidance

`onCallNotPermitted` reports when calls are rejected, because the bulkhead is full.

Reply:
onCallNotPermitted will be covered by the internal semaphore bulkhead events already !, covered the rest except the state events as we do not have implicit states in our implementations